flowexperimental/comp/wells/CompWell_impl.hpp

```diff
@@ -0,0 +1,685 @@
+                      const Well::InjectionControls& inj_controls,
+                      EvalWell& control_eq) const
+{
+    // TODO: we only need to pass in the current control?
+    const auto current = well_state.injection_cmode;
+
+    switch (current) {
+    case WellInjectorCMode::BHP : {
+        const Scalar bhp_limit = inj_controls.bhp_limit;
+        control_eq = this->primary_variables_.getBhp() - bhp_limit;
+        break;
+    }
+    case WellInjectorCMode::RATE : {
+        const Scalar rate_target = inj_controls.surface_rate;
+        const EvalWell& injection_rate = this->primary_variables_.getTotalRate();
+        control_eq = injection_rate - rate_target;
+        break;
+    }
+    default:
+        OPM_THROW(std::logic_error, "only handles BHP and RATE control for injectors for now");
+    }
+}
+
+
+template <typename TypeTag>
+void
+CompWell<TypeTag>::
+assembleSourceTerm(const Scalar dt)
+{
+    // calculating the injection mass rate for each component
+    const EvalWell total_surface_rate = this->primary_variables_.getTotalRate();
+    const EvalWell density = this->surface_conditions_.density();
+    const EvalWell total_mass_rate = total_surface_rate * density;
+    std::array<EvalWell, FluidSystem::numComponents> component_mass_rates;
+    for (unsigned  comp_idx = 0; comp_idx < FluidSystem::numComponents; ++comp_idx) {
+        component_mass_rates[comp_idx] = total_mass_rate * this->surface_conditions_.massFraction(comp_idx);
+    }
+
+    for (unsigned comp_idx = 0; comp_idx < FluidSystem::numComponents; ++comp_idx) {
+        const EvalWell residual = (this->new_component_masses_[comp_idx] - this->component_masses_[comp_idx]) / dt - component_mass_rates[comp_idx];
+        // let us put it in the well equation
+        for (int pvIdx = 0; pvIdx < PrimaryVariables::numWellEq; ++pvIdx) {
+            this->well_equations_.D()[0][0][comp_idx][pvIdx] += residual.derivative(pvIdx + PrimaryVariables::numResEq);
+        }
+        this->well_equations_.residual()[0][comp_idx] += residual.value();
+    }
+}
+
+template <typename TypeTag>
+bool
+CompWell<TypeTag>::
+iterateWellEq(const Simulator& simulator,
+              const Scalar dt,
+              SingleCompWellState<Scalar>& well_state)
+{
+    constexpr int max_iter = 200;
+
+    int it = 0;
+    bool converged = false;
+
+    do {
+        updateWellControl(simulator.vanguard().summaryState(), well_state);
+
+        assembleWellEq(simulator, well_state, dt);
+
+        // get convergence
+        converged = this->getConvergence();
+
+        if (converged) {
+#if 0
+            std::cout << " the well " << this->well_ecl_.name() << " has converged after " << it << " iterations" << std::endl;
+            std::cout << " the residuals ";
+            for (const auto& val : this->well_equations_.residual()[0]) {
+                std::cout << val << " ";
+            }
+            std::cout << std::endl;
+#endif
```

Comment:
Can we not add debugging code into the main sources, please?

Reply:
It is removed and another occasion is replaced with OpmLog::info(), hopefully it will be okay.